docs/design/features/InlineArrayAttribute.md

```diff
@@ -0,0 +1,95 @@
+### Background and motivation
+
+There were multiple proposals and ideas in the past asking for a no-indirection primitive for inline data. [Example1(inline strings)](https://github.com/dotnet/csharplang/issues/2099),  [Example2(inline arrays)](https://github.com/dotnet/runtime/issues/12320), [Example3(generalized fixed-buffers)](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/low-level-struct-improvements.md#safe-fixed-size-buffers)
+Our preexisting offering in this area – [unsafe fixed-sized buffers](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/unsafe-code#fixed-size-buffers) has multiple constraints, in particular it works only with blittable value types and provides no overrun/type safety, which considerably limits its use.
+
+*The InlineArrayAttribute is a building block to allow efficient, type-safe, overrun-safe indexable/sliceable inline data.*
+
+An example of one such feature is [`Inline Arrays` in C# 12.0](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/inline-arrays.md).
+
+### API
+
+```C#
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Struct, AllowMultiple = false)]
+    public sealed class InlineArrayAttribute : Attribute
+    {
+        public InlineArrayAttribute (int length)
+        {
+            Length = length;
+        }
+
+        public int Length { get; }
+    }
+}
+```
+
+When `InlineArray` attribute is applied to a struct with one instance field, it is interpreted by the runtime as a directive to replicate the layout of the struct `Length` times. That includes replicating GC tracking information if the struct happens to contain managed pointers.
+
+Unlike "filed0; field1; field2;..." approach, the resulting layout would be guaranteed to have the same order and packing details as elements of an array with element `[0]` matching the location of the single specified field. That will allow the whole aggregate to be safely indexable/sliceable.
+
+`Length` must be greater than 0.
+
+struct must not have explicit layout.
```

Comment:
What is the behavior if the struct has `Auto` layout? Is it guaranteed the same as sequential and that it will be exactly `sizeof(T) * Length` (where `T` is the type of the singular instance field) long with all data starting at offset 0?

Reply:
It should be the same for `Auto` layout. Since we require exactly one instance field, there is not much difference between `Sequential` and `Auto` for the first element. 
If there are more elements, they start at `sizeof(T) * n` offsets.